core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java

```diff
@@ -291,51 +308,180 @@ public void requestSpace(long sessionId, long blockId, long additionalBytes)
-        additionalBytes, FREE_SPACE_TIMEOUT_MS, blockId);
   }
 
   @Override
-  public void moveBlock(long sessionId, long blockId, BlockStoreLocation newLocation)
+  public void moveBlock(long sessionId, long blockId, AllocateOptions moveOptions)
       throws BlockDoesNotExistException, BlockAlreadyExistsException, InvalidWorkerStateException,
       WorkerOutOfSpaceException, IOException {
-    moveBlock(sessionId, blockId, BlockStoreLocation.anyTier(), newLocation);
+    moveBlock(sessionId, blockId, BlockStoreLocation.anyTier(), moveOptions);
   }
 
   @Override
   public void moveBlock(long sessionId, long blockId, BlockStoreLocation oldLocation,
-      BlockStoreLocation newLocation)
+      AllocateOptions moveOptions)
           throws BlockDoesNotExistException, BlockAlreadyExistsException,
           InvalidWorkerStateException, WorkerOutOfSpaceException, IOException {
-    LOG.debug("moveBlock: sessionId={}, blockId={}, oldLocation={}, newLocation={}", sessionId,
-        blockId, oldLocation, newLocation);
-    RetryPolicy retryPolicy = new TimeoutRetry(FREE_SPACE_TIMEOUT_MS, EVICTION_INTERVAL_MS);
-    while (retryPolicy.attempt()) {
-      MoveBlockResult result = moveBlockInternal(sessionId, blockId, oldLocation, newLocation);
-      if (result.getSuccess()) {
-        synchronized (mBlockStoreEventListeners) {
-          for (BlockStoreEventListener listener : mBlockStoreEventListeners) {
-            listener.onMoveBlockByClient(sessionId, blockId, result.getSrcLocation(),
-                result.getDstLocation());
-          }
+    LOG.debug("moveBlock: sessionId={}, blockId={}, oldLocation={}, options={}", sessionId,
+        blockId, oldLocation, moveOptions);
+    MoveBlockResult result = moveBlockInternal(sessionId, blockId, oldLocation, moveOptions);
+    if (result.getSuccess()) {
+      synchronized (mBlockStoreEventListeners) {
+        for (BlockStoreEventListener listener : mBlockStoreEventListeners) {
+          listener.onMoveBlockByClient(sessionId, blockId, result.getSrcLocation(),
+              result.getDstLocation());
         }
-        return;
       }
+      return;
     }
     // TODO(bin): We are probably seeing a rare transient failure, maybe define and throw some
     // other types of exception to indicate this case.
-    throw new WorkerOutOfSpaceException(ExceptionMessage.NO_SPACE_FOR_BLOCK_MOVE_TIMEOUT,
-        newLocation, blockId, FREE_SPACE_TIMEOUT_MS);
+    throw new WorkerOutOfSpaceException(ExceptionMessage.NO_SPACE_FOR_BLOCK_MOVE,
+        moveOptions.getLocation(), blockId);
+  }
+
+  @Override
+  public void swapBlocks(long sessionId, long blockId1, long blockId2)
+      throws BlockDoesNotExistException, BlockAlreadyExistsException, InvalidWorkerStateException,
+      WorkerOutOfSpaceException, IOException {
+    BlockMeta blockMeta1;
+    BlockMeta blockMeta2;
+    boolean doFsSwap = false;
+
+    long lockId1 = mLockManager.lockBlock(sessionId, blockId1, BlockLockType.READ);
```

Comment:
should there be a log debug message, like the other methods?

Reply:
Api is no more